base/compiler/abstractinterpretation.jl

```diff
@@ -164,13 +189,47 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
             is_improvable(rettype) && InferenceParams(interp).ipo_constant_propagation
         # if there's a possibility we could constant-propagate a better result
         # (hopefully without doing too much work), try to do that now
-        # TODO: it feels like this could be better integrated into abstract_call_method / typeinf_edge
-        const_rettype = abstract_call_method_with_const_args(interp, rettype, f, argtypes, applicable[nonbot]::MethodMatch, sv, edgecycle)
-        if const_rettype ⊑ rettype
-            # use the better result, if it's a refinement of rettype
+        # TODO: refactor this, enable constant propagation for each (union-split) signature
+        match = applicable[nonbot]::MethodMatch
+        const_rettype = abstract_call_method_with_const_args(interp, rettype, f, argtypes, match, sv, edgecycle)
+        const_conditional = ignorelimited(const_rettype)
+        @assert !(const_conditional isa Conditional) "invalid lattice element returned from inter-procedural context"
+        const_rettype = widenwrappedconditional(const_rettype)
+        if ignorelimited(const_rettype) ⊑ rettype
+            # use the better result, if it is a refinement of rettype
             rettype = const_rettype
+            if const_conditional isa InterConditional && conditionals === nothing && fargs !== nothing
+                arg = fargs[const_conditional.slot]
+                if arg isa Slot
+                    rettype = Conditional(arg, const_conditional.vtype, const_conditional.elsetype)
+                    if const_rettype isa LimitedAccuracy
+                        rettype = LimitedAccuracy(rettype, const_rettype.causes)
+                    end
+                end
```

Comment:
Oh, previous `rettype = Conditional(SlotNumber(const_conditional.slot), const_conditional.vtype, const_conditional.elsetype)` was just wrong. Thanks for the fix.